docs/source/scheduling-policies.rst

```diff
@@ -187,9 +207,10 @@ scheduler, and workers at various points in the computation.
     policy between computations.  All tasks from a previous call to compute
     have a higher priority than all tasks from a subsequent call to compute (or
     submit, persist, map, or any operation that generates futures).
-3.  Whenever a task is ready to run the scheduler assigns it to a worker.  The
-    scheduler does not wait based on priority.
-4.  However when the worker receives these tasks it considers their priorities
-    when determining which tasks to prioritize for communication or for
+3.  Whenever a task is ready to run (its dependencies, if any, are complete),
+    the scheduler assigns it to a worker. When multiple tasks are ready at once,
+    the order in which they are scheduled is undefined, not based on priority.
```

Comment:
this doesn't sound right. I understand that the scheduler should stop sending tasks to the workers once the task pressure on them is enough, and that there should still be a depth-first priority among ready tasks?

Reply:
Your intuition that this doesn't sound right is correct, because it's not good scheduling logic. However, it's right in that it's what actually happens today (second half is wrong actually, see below). 

> I understand that the scheduler should stop sending tasks to the workers once the task pressure on them is enough

If this happened, we wouldn't have root task overproduction. In fact, this is one of the solutions I suggested at the bottom of https://github.com/dask/distributed/issues/5223 (don't tell workers about extra tasks in the first place).

The criteria for a task to transition to `processing` is only that it's not waiting on any dependencies. (There's no throttling of the number of tasks that can be in `processing` at once; it can be far more than the workers actually have capacity for.) `transition_waiting_processing` is the only path for a task to get scheduled (besides stealing). `transition_waiting_processing` will only _not_ schedule the task if `decide_worker` returns None: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L2737-L2739
`decide_worker` [only returns None](https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L2613-L2622) if there are no valid workers for the task, which is just a question of worker/host/resource restrictions. There's no logic checking whether the task pressure on the workers is too high.

Because this backpressure doesn't exist, the scheduler oversaturates workers with root tasks, and inevitable race conditions cause them to run too many of them, causing root task overproduction.

> there should still be a depth-first priority among ready tasks

```suggestion
    they are all submitted to workers, in priority order.
```

Reading more carefully, my original changes here were wrong, and we can actually say this ^. I initially said "the order in which they are scheduled is undefined" because `recommendations` aren't pre-sorted during `transitions`, so there's no mechanism guaranteeing or validating that they're scheduled in priority order. However:
1. When tasks initially get added via `update_graph`, the recommendations are added in priority order: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L4875-L4877
2. When a task completes, its dependents are recommended in priority order: https://github.com/dask/distributed/blob/39c5e8856cb05ad1833a0f6ee6c3fea0ddab1c1c/distributed/scheduler.py#L8151-L8153

So in practice, tasks are assigned to workers in priority order. Indeed, https://github.com/dask/distributed/pull/4967 relies on this fact.

However, it's important to note the "all" here. Yes, `decide_worker` does end up being called on the tasks in priority order, but because every runnable task is going to get scheduled in that cycle anyway, it's not that significant from a scheduling perspective what order `decide_worker` gets called on them. If there were backpressure, and after some cutoff we didn't schedule any more, then priority order would matter a lot more, since it would decide which tasks make it in before the cutoff.